MapboxNavigation/NavigationMapView.swift

```diff
@@ -1350,3 +1357,114 @@ extension NavigationMapView {
+            highlightedBuildingLayer?.sourceLayerIdentifier = "building"
+            highlightedBuildingLayer?.fillExtrusionColor = NSExpression(forConstantValue: UIColor.white)
+            highlightedBuildingLayer?.fillExtrusionHeightTransition = MGLTransition(duration: 0.8, delay: 0)
+            highlightedBuildingLayer?.fillExtrusionOpacityTransition = MGLTransition(duration: 0.8, delay: 0)
+  
+            if let highlightedBuildingLayer = highlightedBuildingLayer {
+                style?.addLayer(highlightedBuildingLayer)
+            }
+        }
+    }
+
+    private func buildingId(coordinate: CLLocationCoordinate2D) -> Int? {
+        let screenCoordinate = convert(coordinate, toPointTo: self)
+        let features = visibleFeatures(at: screenCoordinate)
+        
+        if let feature = features.first, let buildingId = feature.identifier as? Int {
+            return buildingId
+        }
+        
+        return nil
+    }
+    
+    private func extrudeBuildings(_ buildings: [(Int, UIColor)], extrudeAll: Bool) {
+        guard buildings.count > 0 else {
+            return
+        }
+        
+        addStyleLayersIfNecessary() // may need to re-add the layers if the map style has changed (e.g. preview style -> nav style)
+        
+        if let highlightedBuildingFillExtrusionLayer = style?.layer(withIdentifier: StyleLayerIdentifier.buildingExtrusion) as? MGLFillExtrusionStyleLayer {
+            var predicate = "extrude = 'true' && type = 'building' && underground = 'false'"
+            if extrudeAll == false {
+                // form a predicate to filter out the other buildings from the datasource so only the desired ones are included
+                var identifiersBoolean = "($featureIdentifier == \(buildings.first!.0)"
+                
+                if buildings.count > 1 {
+                    for building in buildings[1...buildings.count-1] {
+                        identifiersBoolean += " || $featureIdentifier == \(building.0)"
+                    }
+                }
+                
+                identifiersBoolean += ")"
+                
+                predicate += " && \(identifiersBoolean)"
+            }
+            
+            highlightedBuildingFillExtrusionLayer.predicate = NSPredicate(format: predicate)
+            
+            // we have some specific building IDs that we will be the highlight color
+            // the rest of the buildings will be extruded but kept at a uniform color
+            var colorsList = [UIColor]()
+            var highlightedBuildingHeightExpression = "MGL_MATCH($featureIdentifier, "
+            var highlightedBuildingColorExpression = "MGL_MATCH($featureIdentifier, "
+            
+            for building in buildings {
+                highlightedBuildingHeightExpression += "\(building.0), height, "
+                highlightedBuildingColorExpression += "\(building.0), %@, "
+                colorsList.append(building.1)
+            }
```

Comment:
We can simplify these expressions based on an assumption that there’s only two sets of buildings: the buildings we want to highlight and everything else.